symphony-bdk-core/src/main/java/com/symphony/bdk/core/service/message/util/MessageParser.java

```diff
@@ -0,0 +1,138 @@
+   *
+   * @param message incoming V4 message to be parsed
+   * @return list of hashtags contained in the message
+   */
+  public static List<String> getHashtags(V4Message message) throws MessageParserException {
+    return getTags(message, EntityTypeEnum.HASHTAG);
+  }
+
+  /**
+   * Parse data inside the message and returns a list containing the text of the cashtags found
+   *
+   * @param message incoming V4 message to be parsed
+   * @return list of cashtags contained in the message
+   */
+  public static List<String> getCashtags(V4Message message) throws MessageParserException {
+    return getTags(message, EntityTypeEnum.CASHTAG);
+  }
+
+  /**
+   * Parse data inside the message and returns a map containing the list of emojis found.
+   * Key of the map are the annotation used to identify the emoji and the values are the their unicode.
+   *
+   * @param message incoming V4 message to be parsed
+   * @return map of emojis contained in the message
+   */
+  public static Map<String, String> getEmojis(V4Message message) throws MessageParserException {
+    return getEmojisList(message);
+  }
+
+  private static List<String> getTags(V4Message message, EntityTypeEnum type) throws MessageParserException {
+    String data = message.getData();
+    if (data == null || data.isEmpty()) {
+      log.info("Data payload of message with id {} is empty, no entities found.", message.getMessageId());
+      return Collections.emptyList();
+    }
+    JsonNode dataJson = getJsonNode(data);
+    List<String> tags = new ArrayList<>();
+    for (JsonNode node : dataJson) {
+      if (isType(type, node)) {
+        tags.add(node.get("id").get(0).get("value").asText());
+      }
+    }
+    return tags;
+  }
+
+  private static Map<String, String> getEmojisList(V4Message message) throws MessageParserException {
+    String data = message.getData();
+    if (data == null || data.isEmpty()) {
+      log.info("Data payload of message with id {} is empty, no entities found.", message.getMessageId());
+      return Collections.emptyMap();
+    }
+    JsonNode dataJson = getJsonNode(data);
+    Map<String, String> emojisMap = new HashMap<>();
+    for (JsonNode node : dataJson) {
+      if (isType(EntityTypeEnum.EMOJI, node)) {
+        emojisMap.put(node.get("data").get("annotation").asText(), node.get("data").get("unicode").asText());
+      }
+    }
+    return emojisMap;
+  }
+
+  private static JsonNode getJsonNode(String data) throws MessageParserException {
+    JsonNode dataJson;
+    try {
+      dataJson = MAPPER.readTree(data);
+    } catch (JsonProcessingException e) {
+      throw new MessageParserException("Failed to extract payload from message data", e);
+    }
+    return dataJson;
+  }
+
+  private static boolean isType(EntityTypeEnum type, JsonNode node) {
+    return type.getValue().equals(node.get("type").asText());
+  }
+
+  @API(status = API.Status.EXPERIMENTAL)
+  public enum EntityTypeEnum {
```

Comment:
Is it something we really want to expose outside of the `MessageParser`?

Reply:
yep, let's make it private maybe

Reply:
making it private and with  @API(status = API.Status.INTERNAL)